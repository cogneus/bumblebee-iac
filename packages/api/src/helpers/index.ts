export * from './get-client-options'
export * from './check-permissions'
export * from './check-api-environment-vars'
export * from './try-parse-value'
