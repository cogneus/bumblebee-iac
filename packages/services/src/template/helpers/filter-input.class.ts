import { QueryCommandInput } from '@aws-sdk/lib-dynamodb'

export type Operator = 'begins_with' | 'equals'
const exprJoin = ' and '

const arrayValues = (array: string[], key: string): Partial<QueryCommandInput> =>
  array.length ? { [key]: array.join(exprJoin) } : {}

const makeExpression = (name: string, value: string, operator: Operator): string => {
  switch (operator) {
    case 'begins_with':
      return `begins_with(${name}, ${value})`
    default:
      return `(${name} = ${value})`
  }
}

export class FilterInput {
  private names: Record<string, string> = {}
  private values: Record<string, string> = {}
  private keyConditions: string[] = []
  private filters: string[] = []

  public addFilter(
    name: string,
    value: string | undefined,
    operator: Operator = 'equals'
  ): FilterInput {
    return this.add(name, value, operator, this.filters)
  }

  public addKeyCondition(
    name: string,
    value: string | undefined,
    operator: Operator = 'equals'
  ): FilterInput {
    return this.add(name, value, operator, this.keyConditions)
  }

  public toQueryInput(): Partial<QueryCommandInput> {
    return Object.keys(this.names).length
      ? {
          ExpressionAttributeNames: this.names,
          ExpressionAttributeValues: this.values,
          ...arrayValues(this.keyConditions, 'KeyConditionExpression'),
          ...arrayValues(this.filters, 'FilterExpression'),
        }
      : {}
  }

  private add(
    name: string,
    value: string | undefined,
    operator: Operator = 'equals',
    array: string[]
  ): FilterInput {
    if (value) {
      const nameKey = `#${name}`
      const valueKey = `:${name}`
      this.names[nameKey] = name
      this.values[valueKey] = value
      array.push(makeExpression(nameKey, valueKey, operator))
    }
    return this
  }
}
