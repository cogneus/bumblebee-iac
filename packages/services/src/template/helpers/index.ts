export * from './keys.helper'
export * from './validate-template.helper'
export * from './generate-id'
export * from './filter-input.class'
