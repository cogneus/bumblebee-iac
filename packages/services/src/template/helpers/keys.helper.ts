import { DBTemplateItem, TemplateVersion } from '../../types'

export const prefixPK = 'template'
export const prefixSK = 'version'
export const prefixGroup = 'group'
export const prefixModified = 'modified'
export const prefixCurrent = 'current'

export const makePK = ({ id }: { id: string }) => `${prefixPK}#${id}`

export const makeSK = ({ version }: { version: number }) =>
  `${prefixSK}#${version.toString().padStart(10, '0')}`

export const makePKGSIGroup = ({ group }: { group: string }) => `${prefixGroup}#${group}`

export const makeSKGSIModified = ({ modified }: TemplateVersion) => `${prefixModified}#${modified}`

export const makeSKCurrent = () => `${prefixCurrent}`

const keyAttrs = ['PK', 'SK', 'PKGSI', 'SKGSI']

export const removeKeyInfo = (dbItem: DBTemplateItem): TemplateVersion =>
  Object.keys(dbItem).reduce<TemplateVersion>(
    (item, key) =>
      keyAttrs.includes(key)
        ? item
        : {
            ...item,
            [key]: dbItem[key],
          },
    {} as TemplateVersion
  )
