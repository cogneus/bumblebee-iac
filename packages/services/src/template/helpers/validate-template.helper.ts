import { EmptyRequest, MissingAttributes } from '../../exceptions'
import { TemplateItem } from '../../types'

export const mandatoryKeys = ['group', 'source', 'target', 'template']

export const validateTemplate = (templateItem: TemplateItem | undefined | null) => {
  if (!templateItem) {
    throw EmptyRequest()
  }

  const missing = mandatoryKeys.filter((key) => {
    const value = templateItem[key]
    return (
      value === null || value === undefined || (typeof value === 'string' && value.length === 0)
    )
  })
  if (missing.length) {
    throw MissingAttributes(missing)
  }
}
