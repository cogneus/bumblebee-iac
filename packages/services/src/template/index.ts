export * from './add'
export * from './get'
export * from './query'
export * from './remove'
export * from './update'
export * from './versions'
