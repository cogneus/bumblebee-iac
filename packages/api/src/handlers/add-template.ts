import { apiGatewayWrapper, warmupWrapper, Token, standardEnvVars } from 'bumblebee-common'
import { TemplateItem, TemplateVersion, addTemplate } from 'bumblebee-services'
import { checkAPIEnvironmentVars, checkWritePermissions, getClientOptions } from '../helpers'
import { TemplatesParams } from '../types'

checkAPIEnvironmentVars()

export const main = warmupWrapper(
  apiGatewayWrapper<TemplateItem, TemplatesParams, Token, TemplateVersion>({
    handler: ({ payload, serviceContext }) => addTemplate(serviceContext, payload),
    ...standardEnvVars(),
    serviceName: 'template-api',
    logName: 'add-template',
    getClientOptions,
    authCheck: checkWritePermissions,
  })
)
