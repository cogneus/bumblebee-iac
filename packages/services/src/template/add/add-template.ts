import { ServiceContext, ensureClient, IDBClient } from 'bumblebee-common'
import { TemplateItem, TemplateVersion } from '../../types'
import { validateTemplate } from '../helpers'
import { createCommands } from './create-commands'
import { generateId } from '../helpers'

export const addTemplate = async (
  context: ServiceContext,
  template: TemplateItem
): Promise<TemplateVersion> => {
  validateTemplate(template)

  const db = ensureClient<IDBClient>(context, 'db')
  const date = new Date().toISOString()
  const newVersion: TemplateVersion = {
    ...template,
    id: generateId(),
    version: 1,
    modified: date,
    created: date,
  }
  const params = createCommands(newVersion, db.tables.templates.name)
  await db.transactWrite(params)
  return newVersion
}
