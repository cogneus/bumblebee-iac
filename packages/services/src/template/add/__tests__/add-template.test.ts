import { serviceContext } from '../../__mocks__/service-context.mock'
import { templateItem } from '../../__mocks__/template.mock'
import { addTemplate } from '../add-template'
import { generateId } from '../../helpers/generate-id'
import { NIL as emptyId } from 'uuid'

jest.mock('../../helpers/generate-id')

describe('Add template tests', () => {
  beforeEach(() => {
    jest.clearAllMocks()
    jest.useFakeTimers()
    jest.setSystemTime(new Date('2023-01-01T00:00:00Z'))
  })

  test('db request should match snapshot', async () => {
    let input
    ;(serviceContext.clients.db?.transactWrite as jest.Mock).mockImplementationOnce((params) => {
      input = params
      return null
    })
    ;(generateId as jest.Mock).mockReturnValue(emptyId)
    const result = await addTemplate(serviceContext, templateItem)
    expect({
      input,
      result,
    }).toMatchSnapshot()
  })
})
