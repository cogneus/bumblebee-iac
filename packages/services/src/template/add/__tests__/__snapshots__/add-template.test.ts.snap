// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add template tests db request should match snapshot 1`] = `
{
  "input": [
    {
      "Put": {
        "ConditionExpression": "attribute_not_exists(#pk) and attribute_not_exists(#sk)",
        "ExpressionAttributeNames": {
          "#pk": "PK",
          "#sk": "SK",
        },
        "Item": {
          "PK": "template#00000000-0000-0000-0000-000000000000",
          "PKGSI": "group#mock-group",
          "SK": "current",
          "SKGSI": "modified#2023-01-01T00:00:00.000Z",
          "created": "2023-01-01T00:00:00.000Z",
          "group": "mock-group",
          "id": "00000000-0000-0000-0000-000000000000",
          "modified": "2023-01-01T00:00:00.000Z",
          "source": "mock-source",
          "target": "mock-target",
          "template": {},
          "version": 1,
        },
        "TableName": "name-templates",
      },
    },
    {
      "Put": {
        "ConditionExpression": "attribute_not_exists(#pk) and attribute_not_exists(#sk)",
        "ExpressionAttributeNames": {
          "#pk": "PK",
          "#sk": "SK",
        },
        "Item": {
          "PK": "template#00000000-0000-0000-0000-000000000000",
          "PKGSI": "template#00000000-0000-0000-0000-000000000000",
          "SK": "version#0000000001",
          "SKGSI": "modified#2023-01-01T00:00:00.000Z",
          "created": "2023-01-01T00:00:00.000Z",
          "group": "mock-group",
          "id": "00000000-0000-0000-0000-000000000000",
          "modified": "2023-01-01T00:00:00.000Z",
          "source": "mock-source",
          "target": "mock-target",
          "template": {},
          "version": 1,
        },
        "TableName": "name-templates",
      },
    },
  ],
  "result": {
    "created": "2023-01-01T00:00:00.000Z",
    "group": "mock-group",
    "id": "00000000-0000-0000-0000-000000000000",
    "modified": "2023-01-01T00:00:00.000Z",
    "source": "mock-source",
    "target": "mock-target",
    "template": {},
    "version": 1,
  },
}
`;
